From 2113e5224fa9e4d719489857aea0cf5f28c1f446 Mon Sep 17 00:00:00 2001
From: Mahesh Kurapati <mahesh.kurapati@amd.com>
Date: Wed, 15 Nov 2023 17:54:13 -0600
Subject: [PATCH] bmcweb: add redfish support for retimer bundle update

The retimer bundle information is available at the uri
/redfish/v1/UpdateService/FirmwareInventory/retimer_bundle_active uri. This
commit has the necessary changes to support it.

Signed-off-by: Mahesh Kurapati <mahesh.kurapati@amd.com>
---
 redfish-core/include/utils/fw_utils.hpp | 176 ++++++++++++++++++++++++
 redfish-core/lib/update_service.hpp     |   4 +
 2 files changed, 180 insertions(+)

diff --git a/redfish-core/include/utils/fw_utils.hpp b/redfish-core/include/utils/fw_utils.hpp
index ca43aeda..6dd6883d 100644
--- a/redfish-core/include/utils/fw_utils.hpp
+++ b/redfish-core/include/utils/fw_utils.hpp
@@ -548,6 +548,182 @@ inline void getVRBundleFw(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
         "org.freedesktop.DBus.Properties", "GetAll",
         "xyz.openbmc_project.Software.BundleVersion");
 }
+/**
+ * @brief Put status of input swId into json response
+ *
+ * This function will put the appropriate Redfish state of the input
+ * Retimer Bundle firmware info to "RetimerBundle": ["FirmwareID", "FirmwareVersion"
+ * "HealthStatus", "Riser", "SlaveAddress"] within the json response
+ *
+ * @param[i,o] aResp    Async response object
+ * @param[i]   swId     The software ID to get status for
+ * @param[i]   dbusSvc  The dbus service implementing the software object
+ *
+ * @return void
+ */
+inline void getRetimerBundleFw(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                        const std::shared_ptr<std::string>& swId,
+                        const std::string& dbusSvc)
+{
+    crow::connections::systemBus->async_method_call(
+        [asyncResp,
+         swId](const boost::system::error_code errorCode,
+               const boost::container::flat_map< std::string,
+                  std::variant<std::string, std::vector<std::string>>>& propertiesList) {
+            if (errorCode)
+            {
+                messages::internalError(asyncResp->res);
+                return;
+            }
+            std::vector<std::string> FirmwareIDs;
+            std::vector<std::string> Risers;
+            std::vector<std::string> SlaveAddress;
+            std::vector<std::string> Status;
+            std::vector<std::string> Versions;
+            std::vector<std::string> Checksum;
+
+            boost::container::flat_map< std::string,
+                   std::variant<std::string, std::vector<std::string>>>::const_iterator it = propertiesList.find("FirmwareID");
+            // first check if required property exist
+            if (it == propertiesList.end())
+            {
+                BMCWEB_LOG_ERROR << "Can't find property \"FirmwareID\"!";
+                messages::propertyMissing(asyncResp->res, "FirmwareID");
+                return;
+            }
+            it = propertiesList.find("Riser");
+            if (it == propertiesList.end())
+            {
+                BMCWEB_LOG_ERROR << "Can't find property \"Riser\"!";
+                messages::propertyMissing(asyncResp->res, "Riser");
+                return;
+            }
+            it = propertiesList.find("SlaveAddress");
+            if (it == propertiesList.end())
+            {
+                BMCWEB_LOG_ERROR << "Can't find property \"SlaveAddress\"!";
+                messages::propertyMissing(asyncResp->res, "SlaveAddress");
+                return;
+            }
+            it = propertiesList.find("Status");
+            if (it == propertiesList.end())
+            {
+                BMCWEB_LOG_ERROR << "Can't find property \"Status\"!";
+                messages::propertyMissing(asyncResp->res, "Status");
+                return;
+            }
+            it = propertiesList.find("Versions");
+            if (it == propertiesList.end())
+            {
+                BMCWEB_LOG_ERROR << "Can't find property \"Versions\"!";
+                messages::propertyMissing(asyncResp->res, "Versions");
+                return;
+            }
+            it = propertiesList.find("Checksum");
+            if (it == propertiesList.end())
+            {
+                BMCWEB_LOG_ERROR << "Can't find property \"Checksum\"!";
+                messages::propertyMissing(asyncResp->res, "Checksum");
+                return;
+            }
+
+            //retrieve all the values
+            for (const auto& propertyPair : propertiesList)
+            {
+               if (propertyPair.first == "FirmwareID")
+               {
+                  const std::vector<std::string>* fwvalue =
+                                std::get_if<std::vector<std::string>>(
+                                    &propertyPair.second);
+                  if (fwvalue == nullptr)
+                  {
+                     return;
+                  }
+                  FirmwareIDs = *fwvalue;
+               }
+               else if (propertyPair.first == "Riser")
+               {
+                   const std::vector<std::string>* riservalue =
+                                 std::get_if<std::vector<std::string>>(
+                                     &propertyPair.second);
+                  if (riservalue == nullptr)
+                  {
+                     return;
+                  }
+                  Risers = *riservalue;
+               }
+               else if (propertyPair.first == "SlaveAddress")
+               {
+                   const std::vector<std::string>* slavevalue =
+                                 std::get_if<std::vector<std::string>>(
+                                     &propertyPair.second);
+                  if (slavevalue == nullptr)
+                  {
+                     return;
+                  }
+                  SlaveAddress = *slavevalue;
+               }
+               else if (propertyPair.first == "Status")
+               {
+                   const std::vector<std::string>* statusvalue =
+                                 std::get_if<std::vector<std::string>>(
+                                     &propertyPair.second);
+                  if (statusvalue == nullptr)
+                  {
+                     return;
+                  }
+                  Status = *statusvalue;
+               }
+               else if (propertyPair.first == "Versions")
+               {
+                   const std::vector<std::string>* vervalue =
+                                 std::get_if<std::vector<std::string>>(
+                                     &propertyPair.second);
+                  if (vervalue == nullptr)
+                  {
+                     return;
+                  }
+                  Versions = *vervalue;
+               }
+               else if (propertyPair.first == "Checksum")
+               {
+                   const std::vector<std::string>* chksumvalue =
+                                 std::get_if<std::vector<std::string>>(
+                                     &propertyPair.second);
+                  if (chksumvalue == nullptr)
+                  {
+                     return;
+                  }
+                  Checksum = *chksumvalue;
+               }
+           }//end of for loop
+           // now process all the property values
+           if (((FirmwareIDs.size() == Risers.size())
+               == (SlaveAddress.size() == Status.size()))
+               == (Versions.size() == Checksum.size()))
+           {
+               for (unsigned i = 0; i < FirmwareIDs.size(); i++ )
+               {
+                   nlohmann::json& members = asyncResp->res.jsonValue["RetimerBundle"];
+                   members.push_back({
+                     {"SlaveAddress", SlaveAddress.at(i)},
+                     {"FirmwareID", FirmwareIDs.at(i)},
+                     {"FirmwareVersion", Versions.at(i)},
+                     {"Riser", Risers.at(i)},
+                     {"HealthStatus", Status.at(i)},
+                     {"Checksum", Checksum.at(i)},
+                  });
+               }
+           }
+           else
+           {
+               BMCWEB_LOG_ERROR << "Unknown Retimer bundle firmware";
+           }
+        },
+        dbusSvc, "/xyz/openbmc_project/software/" + *swId,
+        "org.freedesktop.DBus.Properties", "GetAll",
+        "xyz.openbmc_project.Software.RetimerBundleVersion");
+}

 /**
  * @brief Updates programmable status of input swId into json response
diff --git a/redfish-core/lib/update_service.hpp b/redfish-core/lib/update_service.hpp
index 1d70cded..382210ac 100644
--- a/redfish-core/lib/update_service.hpp
+++ b/redfish-core/lib/update_service.hpp
@@ -862,6 +862,10 @@ class SoftwareInventory : public Node
                         fw_util::getVRBundleFw(asyncResp, swId, obj.second[0].first);
                     }

+                    if (*swId == "retimer_bundle_active")
+                    {
+                        fw_util::getRetimerBundleFw(asyncResp, swId, obj.second[0].first);
+                    }
                     crow::connections::systemBus->async_method_call(
                         [asyncResp,
                          swId](const boost::system::error_code errorCode,
--
2.17.1

